sign/dilithium/internal/field.go

```diff
@@ -0,0 +1,125 @@
+package internal
+
+// Returns a y with y < 2q and y = x mod q.
+// Note that in general *not*: reduceLe2Q(reduceLe2Q(x)) == x.
+func reduceLe2Q(x uint32) uint32 {
+	// Note 2^23 = 2^13 - 1 mod q. So, writing  x = x1 2^23 + x2 with x2 < 2^23
+	// and x1 < 2^9, we have x = y (mod q) where
+	// y = x2 + x1 2^13 - x1 ≤ 2^23 + 2^13 < 2q.
+	x1 := x >> 23
+	x2 := x & 0x7FFFFF // 2^23-1
+	return x2 + (x1 << 13) - x1
+}
+
+// Returns x mod q.
+func modQ(x uint32) uint32 {
+	return le2qModQ(reduceLe2Q(x))
+}
+
+// For x R ≤ q 2^32, find y ≤ 2q with y = x mod q.
+func montReduceLe2Q(x uint64) uint32 {
+	// Qinv = 4236238847 = -(q^-1) mod 2^32
+	m := (x * Qinv) & 0xffffffff
+	return uint32((x + m*uint64(Q)) >> 32)
+}
+
+// Returns x mod q for 0 ≤ x < 2q.
+func le2qModQ(x uint32) uint32 {
+	x -= Q
+	mask := uint32(int32(x) >> 31) // mask is 2^32-1 if x was neg.; 0 otherwise
+	return x + (mask & Q)
+}
+
+// Splits 0 ≤ a < Q into a0 and a1 with a = a1*2^D + a0
+// and -2^{D-1} < a0 < 2^{D-1}.  Returns a0 + Q and a1.
+func power2round(a uint32) (a0plusQ, a1 uint32) {
+	// We effectively compute a0 = a mod± 2^d
+	//                    and a1 = (a - a0) / 2^d.
+	a0 := a & ((1 << D) - 1) // a mod 2^d
+
+	// a0 is one of  0, 1, ..., 2^{d-1}-1, 2^{d-1}, 2^{d-1}+1, ..., 2^d-1
+	a0 -= (1 << (D - 1)) + 1
+	// now a0 is     -2^{d-1}-1, -2^{d-1}, ..., -2, -1, 0, ..., 2^{d-1}-2
+	// Next, we add 2^D to those a0 that are negative (seen as int32).
+	a0 += uint32(int32(a0)>>31) & (1 << D)
+	// now a0 is     2^{d-1}-1, 2^{d-1}, ..., 2^d-2, 2^d-1, 0, ..., 2^{d-1}-2
+	a0 -= (1 << (D - 1)) - 1
+	// now a0 id     0, 1, 2, ..., 2^{d-1}-1, 2^{d-1}-1, -2^{d-1}-1, ...
+	// which is what we want.
+	a0plusQ = Q + a0
+	a1 = (a - a0) >> D
+	return
+}
+
+// Splits 0 ≤ a < Q into a0 and a1 with a = a1*α + a0 with -α/2 < a0 ≤ α/2,
+// except for when we would have a1 = (Q-1)/α = 16 in which case a1=0 is taken
+// and -α/2 ≤ a0 < 0.  Returns a0 + Q.  Note 0 ≤ a1 ≤ 15.
```

Comment:
You might like this https://github.com/mmcloughlin/mathfmt 😄

Reply:
Very cool!